<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Vishnu Raghuram">

<title>Module 3 – NWU_2025_Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-45a48b56c8ad2523a9a31c69be39928e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">NWU_2025_Workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">Bacterial Pathogen Genomics Workshop</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module0.html"> 
<span class="menu-text">Module 0</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module1.html"> 
<span class="menu-text">Module 1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module2.html"> 
<span class="menu-text">Module 2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./module3.html" aria-current="page"> 
<span class="menu-text">Module 3</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module4.html"> 
<span class="menu-text">Module 4</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#public-genomic-databases-for-pathogen-surveillance" id="toc-public-genomic-databases-for-pathogen-surveillance" class="nav-link active" data-scroll-target="#public-genomic-databases-for-pathogen-surveillance">Public Genomic Databases for Pathogen Surveillance</a></li>
  <li><a href="#key-ncbi-resources" id="toc-key-ncbi-resources" class="nav-link" data-scroll-target="#key-ncbi-resources">Key NCBI Resources</a>
  <ul class="collapse">
  <li><a href="#bioproject-biosample-and-sra" id="toc-bioproject-biosample-and-sra" class="nav-link" data-scroll-target="#bioproject-biosample-and-sra">BioProject, BioSample, and SRA</a>
  <ul class="collapse">
  <li><a href="#bioproject" id="toc-bioproject" class="nav-link" data-scroll-target="#bioproject">BioProject</a></li>
  <li><a href="#biosample" id="toc-biosample" class="nav-link" data-scroll-target="#biosample">BioSample</a></li>
  <li><a href="#sra-sequence-read-archive-raw-fastq-reads" id="toc-sra-sequence-read-archive-raw-fastq-reads" class="nav-link" data-scroll-target="#sra-sequence-read-archive-raw-fastq-reads">SRA (Sequence Read Archive): Raw fastq reads</a></li>
  <li><a href="#nucleotide-sequences" id="toc-nucleotide-sequences" class="nav-link" data-scroll-target="#nucleotide-sequences">Nucleotide sequences</a></li>
  </ul></li>
  <li><a href="#retrieving-data-from-ncbi-via-command-line" id="toc-retrieving-data-from-ncbi-via-command-line" class="nav-link" data-scroll-target="#retrieving-data-from-ncbi-via-command-line">Retrieving Data from NCBI via Command Line</a>
  <ul class="collapse">
  <li><a href="#ncbi-datasets-cli" id="toc-ncbi-datasets-cli" class="nav-link" data-scroll-target="#ncbi-datasets-cli">ncbi-datasets CLI</a></li>
  <li><a href="#ncbi-genome-download" id="toc-ncbi-genome-download" class="nav-link" data-scroll-target="#ncbi-genome-download">ncbi-genome-download</a></li>
  <li><a href="#sra-tools" id="toc-sra-tools" class="nav-link" data-scroll-target="#sra-tools">sra-tools</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#exploring-enterobase" id="toc-exploring-enterobase" class="nav-link" data-scroll-target="#exploring-enterobase">Exploring EnteroBase</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Module 3</h1>
<p class="subtitle lead">Using publicly available data for pathogen surveillance</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Vishnu Raghuram </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="public-genomic-databases-for-pathogen-surveillance" class="level1">
<h1>Public Genomic Databases for Pathogen Surveillance</h1>
<ul>
<li>Free, accessible, and global data can be integrated for comprehensive monitoring of infectious diseases</li>
<li>Helps contextualize local outbreaks/epidemics on a global scale (i.e: Has this strain been seen before? When? Where?)</li>
<li>Allows comparisons between different time periods, geographical locations, and strains of different genetic backgrounds.</li>
</ul>
<p><strong>Understanding how to use the different sources of publicly available data along with careful interpretation of metadata is key for pathogen genomics research</strong></p>
</section>
<section id="key-ncbi-resources" class="level1">
<h1>Key NCBI Resources</h1>
<section id="bioproject-biosample-and-sra" class="level2">
<h2 class="anchored" data-anchor-id="bioproject-biosample-and-sra">BioProject, BioSample, and SRA</h2>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="assets/images/module3/NCBI_tree.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>NCBI tree</figcaption>
</figure>
</div>
<section id="bioproject" class="level3">
<h3 class="anchored" data-anchor-id="bioproject">BioProject</h3>
<p>The BioProject serves as a central repository linking all deposited data for a specific research project. The BioProject is a single place that can be used to find all other diverse data types deposited under that one project.</p>
<p>BioProjects will have identifiers (or accession number) starting with <code>PRJNA</code></p>
<p>The BioProject accession for the source of the sample data we are using in this module is <a href="https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA1230142">PRJNA1230142</a>. Click on the link and take a look at the information available.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/PRJNA1230142_bioproject_1.png" class="img-fluid figure-img"></p>
<figcaption>BioProject details</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/PRJNA1230142_bioproject_2.png" class="img-fluid figure-img"></p>
<figcaption>Available data under a single BioProject</figcaption>
</figure>
</div>
<p>Below which, you can also find individual links to each sample record under the BioProject along with the option to download the summary table.</p>
</section>
<section id="biosample" class="level3">
<h3 class="anchored" data-anchor-id="biosample">BioSample</h3>
<p>Like how the BioProject is the link to all samples deposited under a single project, the BioSample is the link to all data deposited for a <strong>single sample</strong>. This includes metadata for a single sample (organism, collection site, etc.), the raw reads for that sample (if deposited), and the genome assembly/annotation data. The BioSample also serves as link back to the BioProject.</p>
<p>BioSamples will have identifiers (or accession number) starting with <code>SAM</code></p>
<p>Click on the BioSample link to any one of the records and take a look.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/BioSample.png" class="img-fluid figure-img"></p>
<figcaption>BioSample details</figcaption>
</figure>
</div>
<p>The <strong>attributes</strong> section in each BioSample record comprises important metadata regarding when and where the sample was collected, and some information about strain background and genotype. Further below you can access the <strong>SRA</strong> (raw reads) and <strong>Nucleotide</strong> (assembly) records for each BioSample.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="The challenge with publicly available metadata">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>The challenge with publicly available metadata
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Remember that metadata is simply what is reported by the submitter. You will often find samples with vague descriptions, typos or even no metadata at all. When you are depositing samples in NCBI make sure to provide rich metadata with no errors!</p>
</div>
</div>
</div>
</section>
<section id="sra-sequence-read-archive-raw-fastq-reads" class="level3">
<h3 class="anchored" data-anchor-id="sra-sequence-read-archive-raw-fastq-reads">SRA (Sequence Read Archive): Raw fastq reads</h3>
<p>SRA is a repository of all high-throughput sequencing data. It contains raw sequence reads from genomic, metagenomic, and transcriptomic sources.</p>
<p>Through the BioSample, you can also access the raw reads by clicking on the <strong>SRA</strong> link at the bottom of the page. This will take you to the SRA <strong>experiment</strong> page.</p>
<p>The SRA experiment has information regarding how the reads were generated (i.e the experimental design). This also links back to the BioProject and BioSample.</p>
<p>The SRA experiment accessions start with <code>SRX</code>, <code>ERX</code> or <code>DRX</code>.</p>
<p>The SRA experiment page also link to the SRA <strong>run</strong> for that sample. The SRA run page has the actual <code>.fastq</code> sequence reads. The SRA runs have accessions that start with <code>SRR</code>,<code>ERR</code> or <code>DRR</code></p>
<p>The SRA run page for a given sample also links back to the SRA experiment, BioSample and BioProject. It also has a download link to the raw <code>fastq</code> reads.</p>
<p>Navigate to the SRR page for any sample and take a look at the information available.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/SRR_download.png" class="img-fluid figure-img"></p>
<figcaption>SRA Run - downloading reads</figcaption>
</figure>
</div>
</section>
<section id="nucleotide-sequences" class="level3">
<h3 class="anchored" data-anchor-id="nucleotide-sequences">Nucleotide sequences</h3>
<p>NCBI has multiple nucleotide sequence databases, but all deposited sequences are stored in <strong>Nuccore</strong> - this can be individual contigs from each assembly, plasmid sequences, or gene sequences.</p>
<p>There are smaller databases with more specific nucleotide sequences branching from Nuccore, for eg: the <strong>Gene</strong> database for gene sequences and <strong>Genome</strong> database for whole genome sequences. The Genome (or ‘Assembly’) database accessions begin with <code>ASM</code>. The <code>ASM</code> record for a given sample provides the assembly itself, assembly statistics, annotation information and also links back to the BioSample and BioProject.</p>
<p>The assembly can be obtained from RefSeq or GenBank, two repositories of genome sequences deposited in NCBI with minor differences.For our purposes, we will mostly be interacting with <strong>GenBank</strong>. GenBank accessions start with <code>GCA</code>. Navigate to the genome assembly page for any sample and take a look.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/ASM.png" class="img-fluid figure-img"></p>
<figcaption>Genome assembly</figcaption>
</figure>
</div>
</section>
</section>
<section id="retrieving-data-from-ncbi-via-command-line" class="level2">
<h2 class="anchored" data-anchor-id="retrieving-data-from-ncbi-via-command-line">Retrieving Data from NCBI via Command Line</h2>
<p>As you may have seen above, it is possible to click through the NCBI website and access various types of information from a given sample. However, this starts becoming impractical as soon as you need files on hundreds or even tens of samples. Luckily, there are useful command-line tools available that let us programmatically retrieve data from NCBI.</p>
<section id="ncbi-datasets-cli" class="level3">
<h3 class="anchored" data-anchor-id="ncbi-datasets-cli">ncbi-datasets CLI</h3>
<p>This is the official NCBI software for accessing different datasets. If you have the BioSample or BioProject accession, <code>datasets</code> lets you download assemblies, annotations and summary information.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">datasets</span> download genome accession GCA_000486855.2 <span class="at">--filename</span> GCA_000486855.2.zip <span class="at">--include</span> cds,genome,protein,gbff,gff3  </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>A <code>zip</code> file will be downloaded which you can extract using the <code>unzip</code> command and access the files.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unzip</span> GCA_000486855.2.zip</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Use <code>datasets download genome --help</code> to see the different options, you can fetch a lot of data by using the right set of options!</p>
</section>
<section id="ncbi-genome-download" class="level3">
<h3 class="anchored" data-anchor-id="ncbi-genome-download">ncbi-genome-download</h3>
<p>This is an alternative to ncbi <code>datasets</code> that is very similar. <code>ncbi-genome-download</code> is not an official tool made and maintained by NCBI, it was just the bioinformatics community’s solution to downloading NCBI datasets before <code>datasets</code> existed. It is still a viable alternative today. It offers some options not found in <code>datasets</code> such as downloading assembly metadata in <code>.tsv</code>.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ncbi-genome-download</span> <span class="at">-A</span> GCA_049745</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">255.1,GCA_049744095.1,GCA_049744075.1,GCA_049744695.1,GCA_049744875.1</span> <span class="at">-m</span> metadata.tsv bacteria</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="sra-tools" class="level3">
<h3 class="anchored" data-anchor-id="sra-tools">sra-tools</h3>
<p>The SRA Toolkit is NCBI’s official software for downloading and extracting <code>fastq</code> files from SRA runs. It has two main parts - <code>prefetch</code> and <code>fasterq-dump</code>.</p>
<p><code>prefetch</code> downloads an SRA run in <code>.sra</code> format which can then be converted to <code>.fastq</code> using <code>fasterq-dump</code>. Alternatively you can also directly download reads using <code>fasterq-dump</code></p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fasterq-dump</span> SRR3252850</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<section id="exercise-retrieve-data-from-ncbi-using-datasets" class="level4">
<h4 class="anchored" data-anchor-id="exercise-retrieve-data-from-ncbi-using-datasets">EXERCISE: retrieve data from NCBI using <code>datasets</code></h4>
<ul>
<li><p>Download only the assemblies in the GenBank database from the BioProject <code>PRJNA1230142</code></p>
<details>
<summary>
<p>Click to reveal answer</p>
</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">datasets</span> download  genome accession PRJNA1230142 <span class="at">--assembly-source</span> <span class="st">'GenBank'</span> <span class="at">--include</span> <span class="st">'genome'</span> </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details></li>
<li><p>Download all “chromosome” and “complete” level <em>Salmonella enterica</em> assemblies and their corresponding annotation files in genbank format from the GenBank database deposited after October 1st, 2025.</p>
<details>
<summary>
<p>Click to reveal answer</p>
</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">datasets</span> download  genome taxon <span class="st">'Salmonella enterica'</span> <span class="at">--assembly-source</span> <span class="st">'GenBank'</span> <span class="at">--assembly-level</span> <span class="st">'chromosome,complete'</span> <span class="at">--include</span> <span class="st">'genome,gbff'</span> <span class="at">--released-after</span> 2025-10-01</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details></li>
</ul>
</section>
</section>
</section>
</section>
<section id="exploring-enterobase" class="level1">
<h1>Exploring EnteroBase</h1>
<p>Enterobase is a web-based platform for genomic epidemiology of key pathogens such as <em>Salmonella</em>, <em>Shigella</em>, MTb, and others.</p>
<p>Like NCBI, EnteroBase is also a repository of sequencing data. However, unlike NCBI, EnteroBase is a lot more specific in its scope. It only hosts data regarding a handful of pathogens, and the focus is more on tracking different lineages temporally and geographically. This makes <strong>EnteroBase</strong> a rich source of metadata for specific organisms.</p>
<p>Head over to https://enterobase.warwick.ac.uk/ and select “Salmonella enterica” and click “Search strains”.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/Enterobase_search.png" class="img-fluid figure-img"></p>
<figcaption>Enterobase Search Filters</figcaption>
</figure>
</div>
<p>Here you can apply several granular filters to access strains with very specific properties, which is not easy to do in NCBI resources.</p>
<p>For example, here is how you can search for only <em>Salmonella enterica</em> serovar Minnesota strains from the African continent isolated from humans</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/Enterobase_search_filters.png" class="img-fluid figure-img"></p>
<figcaption>Enterobase Search Filters</figcaption>
</figure>
</div>
<p>After which you can select and download the assemblies or the data tables (of which there are multiple to choose from). Apart from ‘Strain Metadata’ filters, we can also add filters to these additional data tables in the ‘Experimental Data’ tab for even more granularity.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module3/Enterobase_search_datatables.png" class="img-fluid figure-img"></p>
<figcaption>Enterobase Search Filters</figcaption>
</figure>
</div>
<section id="exercise-retrieve-data-from-enterobase" class="level4">
<h4 class="anchored" data-anchor-id="exercise-retrieve-data-from-enterobase">EXERCISE: Retrieve data from enterobase</h4>
<ul>
<li>Filter for all <em>Salmonella enterica</em> serovar Minnesota genomes from both Africa and Europe collected after the year 2023 (2024 or above) having the Sulfonamide antibiotic resistance gene <em>sul2</em></li>
</ul>
<details>
<summary>
Click to reveal answer
</summary>
<p><img src="assets/images/module3/Enterobase_search_filters_exercise_1.png" class="img-fluid" alt="Filters in the Strain Metadata tab"> <img src="assets/images/module3/Enterobase_search_filters_exercise_2.png" class="img-fluid" alt="Filters in the Experimental Data tab"></p>
</details>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>