<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Vishnu Raghuram &amp; Laura Carroll">

<title>Module 4 – NWU_2025_Workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-45a48b56c8ad2523a9a31c69be39928e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="./index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">NWU_2025_Workshop</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">Bacterial Pathogen Genomics Workshop</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module0.html"> 
<span class="menu-text">Module 0</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module1.html"> 
<span class="menu-text">Module 1</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module2.html"> 
<span class="menu-text">Module 2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./module3.html"> 
<span class="menu-text">Module 3</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./module4.html" aria-current="page"> 
<span class="menu-text">Module 4</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction-to-genomic-distance-estimation" id="toc-introduction-to-genomic-distance-estimation" class="nav-link active" data-scroll-target="#introduction-to-genomic-distance-estimation">Introduction to Genomic Distance Estimation</a>
  <ul class="collapse">
  <li><a href="#multi-locus-sequence-typing-mlst" id="toc-multi-locus-sequence-typing-mlst" class="nav-link" data-scroll-target="#multi-locus-sequence-typing-mlst">Multi-Locus Sequence Typing (MLST)</a></li>
  <li><a href="#minhash-distance-estimation" id="toc-minhash-distance-estimation" class="nav-link" data-scroll-target="#minhash-distance-estimation">MinHash distance estimation</a></li>
  <li><a href="#ani-average-nucleotide-identity" id="toc-ani-average-nucleotide-identity" class="nav-link" data-scroll-target="#ani-average-nucleotide-identity">ANI (Average Nucleotide Identity)</a></li>
  <li><a href="#snp-distances" id="toc-snp-distances" class="nav-link" data-scroll-target="#snp-distances">SNP Distances</a>
  <ul class="collapse">
  <li><a href="#choosing-a-reference-genome" id="toc-choosing-a-reference-genome" class="nav-link" data-scroll-target="#choosing-a-reference-genome">Choosing a reference genome:</a></li>
  <li><a href="#recombination-filtering" id="toc-recombination-filtering" class="nav-link" data-scroll-target="#recombination-filtering">Recombination filtering</a></li>
  <li><a href="#core-genome-vs-whole-genome-alignments" id="toc-core-genome-vs-whole-genome-alignments" class="nav-link" data-scroll-target="#core-genome-vs-whole-genome-alignments">Core genome vs whole genome alignments</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Module 4</h1>
<p class="subtitle lead">Genotyping and genomic distance estimation</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Vishnu Raghuram &amp; Laura Carroll </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="introduction-to-genomic-distance-estimation" class="level1">
<h1>Introduction to Genomic Distance Estimation</h1>
<p>The measure of overall (dis)similarity between two or more genomes helps in understanding relationships between different organisms.</p>
<p>This is often used for typing bacterial strains in epidemiological studies for species-level to strain-level (subspecies) identification.</p>
<p>This can also be used to quickly identify any outliers or unusual genomes in your dataset.</p>
<p>Genomic distance estimation across your dataset will (<em>or should!</em>) be the first thing you do before beginning your analyses.</p>
<p>There are multiple approaches to go about this ranging from fast and approximate to slow and accurate.</p>
<section id="multi-locus-sequence-typing-mlst" class="level2">
<h2 class="anchored" data-anchor-id="multi-locus-sequence-typing-mlst">Multi-Locus Sequence Typing (MLST)</h2>
<ul>
<li><p>MLST is a method used to classify bacteria based on the sequences specific, highly conserved housekeeping genes.</p></li>
<li><p>Since housekeeping genes are highly stable (less prone to random mutations as their functions are critical), this provides a stable and reproducible way of typing bacterial strains.</p></li>
<li><p>From a pathogen surveillance perspective, it is useful for quickly identifying the bacterial strain that is responsible for an ongoing outbreak.</p></li>
<li><p>There are multiple MLST methods, the most common being 7-gene MLST. This method identifies 7 specific housekeeping gene alleles and assigns a type, referred to as the <strong>ST</strong> (Sequence Type), based on the combination of alleles present.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/module4/7GeneMLST.png" class="img-fluid figure-img"></p>
<figcaption>7 Gene MLST</figcaption>
</figure>
</div>
<p>The specific 7 genes used for this typing differs across species. This is referred to as the MLST <strong>schema</strong>. Different organisms are typed according to different MLST schema. As we saw before, EnteroBase has MLST information for certain species. <a href="https://pubmlst.org/">PubMLST</a> is also a good resource to explore different MLST schema for different species.</p>
<section id="exercise-assign-sequence-types-using-mlst" class="level4">
<h4 class="anchored" data-anchor-id="exercise-assign-sequence-types-using-mlst">EXERCISE: Assign Sequence Types Using MLST</h4>
<p>Use the <code>mlst</code> tool to identify the ST type for the sample genomes. They should be available in <code>test_datasets/salmonella_assemblies/</code>, or as we saw in the previous module you can download assemblies using <code>datasets</code>.</p>
<details>
<summary>
Click to reveal answer
</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mlst</span> <span class="at">--nopath</span> NWU_2025_workshop_data/test_datasets/salmonella_assemblies/<span class="pp">*</span>.fna <span class="op">&gt;</span> mlst_table.tsv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><code>NWU_2025_workshop_data/test_datasets/salmonella_assemblies/*.fna</code> : What does the <code>*</code> mean here?</p>
<ul>
<li>The <code>*</code> signifies a <strong>wildcard</strong>.</li>
<li><strong>wildcards</strong> in <code>bash</code> are specific characters that have special meaning and they can help in finding files, directories or text that match patterns, as opposed to having to type out the full names every time.</li>
<li><code>*</code> means “anything” . <code>*.fna</code> means anything that ends with <code>.fna</code>. <code>G*.fna</code> means anything that begins with a <code>G</code> and ends with <code>fna</code>, and so on.</li>
<li>There are several other wildcards in bash with different special meanings.</li>
<li>Wildcards are powerful tools that can help you quickly find text, make specific manipulations to text files, and find files and folders in your computer by specifying patterns.</li>
<li>Learn more about wildcards <a href="https://bashcommands.com/bash-wildcard">here</a>!</li>
</ul>
</details>
<p>Above we used the 7-Gene MLST method, however there are other MLST methods such as wgMLST and cgMLST which instead of only 7 genes, use alleles from the <u>w</u>hole <u>g</u>enome or the <u>c</u>ore <u>g</u>enome respectively. These methods use more than just seven genes and therefore offer higher resolution in typing. <a href="https://chewbbaca.readthedocs.io/en/latest/user/getting_started/overview.html">chewBBACA</a> is a popular software for creating and determining cgMLST/wgMLST.</p>
<p>Depending on the organism, different typing schemes are widely used. Consult up-to-date literature for your organism of interest to see which is the most prevalent typing scheme in your case.</p>
</section>
</section>
<section id="minhash-distance-estimation" class="level2">
<h2 class="anchored" data-anchor-id="minhash-distance-estimation">MinHash distance estimation</h2>
<ul>
<li>MinHash or MASH is a method for estimating approximate genomic distances by comparing <strong>k-mers</strong></li>
<li>A given whole genome sequence is divided into several very short sequences called <strong>K-mers</strong>, where <strong>K</strong> stands for the length of the short sequences. (for example, if the genome was divided into several 31bp long sequences, K = 31)</li>
<li>These K-mers are then compared to see how many identical and how many non-identical k-mers there are between two different genomes.</li>
<li>The exact computation can be seen <a href="https://mash.readthedocs.io/en/latest/distances.html#mash-distance-formulation">here</a> but in effect this gives you an approximate similarity or dissimilarity score.</li>
<li>This method is very fast and computationally efficient, making it easy to compare thousands to tens of thousands of genomes in a very short time.</li>
</ul>
<section id="exercise-calculate-pairwise-mash-distances" class="level4">
<h4 class="anchored" data-anchor-id="exercise-calculate-pairwise-mash-distances">Exercise: Calculate pairwise MASH distances</h4>
<p>Use the <code>mashtree</code> tool to calculate all-vs-all pairwise MASH distances for the genomes in <code>test_datasets/salmonella_assemblies/</code>. “all-vs-all pairwise” means we will be comparing every genome in our set to every other genome.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>mashtree</code> by default only reports a <strong>dendrogram</strong> - which is a tree-like hierarchical grouping of similar items. We need this dendrogram for visualization purposes, but make sure to also output a table or matrix so that we have the actual distance values as well</p>
</div>
</div>
<details>
<summary>
Click to reveal answer
</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mashtree</span> <span class="at">--numcpus</span> 4 <span class="at">--outmatrix</span> matrix.tsv <span class="at">--outtree</span> tree.dnd NWU_2025_workshop_data/test_datasets/salmonella_assemblies/<span class="pp">*</span>.fna</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</details>
</section>
</section>
<section id="ani-average-nucleotide-identity" class="level2">
<h2 class="anchored" data-anchor-id="ani-average-nucleotide-identity">ANI (Average Nucleotide Identity)</h2>
<ul>
<li>A method to calculate the genomic similarity by comparing the percentage of identical nucleotide sequences between two genomes.</li>
<li>It is similar to MASH in that the two genomes being compared are fragmented into shorter sequences but these fragments are much larger (~1000 bp) than the MASH k-mers.</li>
<li>The sequence identity of each fragment to its orthologous fragment (i.e the most similar fragment in the other genome) is measured and this is done for all orthologous fragments.</li>
<li>The average of this nucleotide identity is then reported.</li>
<li>This method is more accurate than MASH but is also more computationally intensive.</li>
<li>ANI is commonly used for species delineation but can also be used for sub-species delineation.</li>
</ul>
<section id="exercise-calculate-pairwise-ani" class="level4">
<h4 class="anchored" data-anchor-id="exercise-calculate-pairwise-ani">EXERCISE: Calculate pairwise ANI</h4>
<p>Use the <code>skani</code> tool to calculate all-vs-all pairwise ANI for the genomes in <code>test_datasets/salmonella_assemblies/</code>.</p>
<p>Specifically you must use <code>skani dist</code> as <code>skani</code> has other functions that we dont need right now. Look at the help page by running <code>skani dist --help</code></p>
<details>
<summary>
Click to reveal answer
</summary>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">skani</span> dist <span class="at">-q</span> NWU_2025_workshop_data/test_datasets/salmonella_assemblies/<span class="pp">*</span>.fna <span class="at">-r</span> NWU_2025_workshop_data/test_datasets/salmonella_assemblies/<span class="pp">*</span>.fna <span class="at">-o</span> skani.tsv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Since we are performing an all-vs-all comparison, the list of query sequences and the list of reference sequences to be compared in our case are the same, hence <code>-q</code> and <code>-r</code> have the same arguments. However with <code>skani</code> (and <code>mash</code>) you can compare two mutually exclusive lists as well, or compare a list with a single genome, depending on your needs.</p>
</details>
<p>Take a look at the ANI values reported by <code>skani</code> in the output for each pair of genomes. How do they compare with the <code>mash</code> distances for the same pair?</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Cleaning up output files">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Cleaning up output files
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You may have noticed that the <code>skani</code> output has the full file paths for each sample (if you specified it in the input). In our final output, this is actually a problem as we cannot match the IDs with the IDs from other results due to the paths being appended. As we learned before, we can use <code>sed</code> to find and replace text, but in this case we have a complex long pattern that changes with each line. In these cases, we can use a powerful concept called <strong>regular expressions</strong>. Regular expressions or <strong>regex</strong> are a way to describe complex patterns as opposed to exact text. The <strong>regex</strong> to remove the paths from the <code>skani</code> output would be:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sed</span> <span class="st">'s/NWU_.*\/\(GCA_.*\)_genomic.fna\tNWU_.*\/\(GCA_.*\)_genomic.fna/\1\t\2/g'</span> skani.tsv <span class="op">&gt;</span> skani_cleaned.tsv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Regular expressions are a complex topic that is beyond the scope of this workshop, but nevertheless it is an essential tool in the command line. You can learn more about regex <a href="https://regexone.com/">here</a>.</p>
</div>
</div>
</div>
</section>
</section>
<section id="snp-distances" class="level2">
<h2 class="anchored" data-anchor-id="snp-distances">SNP Distances</h2>
<p>We learned about SNP (Single Nucleotide Polymorphism) calling in Module2. When performing SNP calling, we align a genome to a reference and identify specific nucleotide differences between the genome and the reference. The exact number of these nucleotide differences is called the SNP distance. This SNP calling can be performed for several genomes, all aligned against the same reference.</p>
<p>and the all-vs-all pairwise number of SNPs can be estimated relative to that reference.</p>
<p>This is the most accurate and the most time consuming way to compare two genomes, however the resolution provided is significantly greater than ANI or MASH. For example, when comparing two genomes that are less than 10 SNPs apart (as is often the case during outbreaks), ANI or MASH will not be able to capture those differences.</p>
<p>However, there are some key factors that need to be considered before calculating SNP distances.</p>
<section id="choosing-a-reference-genome" class="level3">
<h3 class="anchored" data-anchor-id="choosing-a-reference-genome">Choosing a reference genome:</h3>
<p>The reference genome will act as an anchor or the starting point for the alignmnent tool. By definition, sequence alignments will be biased to the reference genome, therefore the choice of reference can greatly impact the quality of the alignment. Ideally, we would want a high quality, complete genome closely related to our samples to act as the reference. This is where the genomic comparison methods we used above can be useful.</p>
<section id="choosing-a-reference-that-is-the-same-mlst" class="level4">
<h4 class="anchored" data-anchor-id="choosing-a-reference-that-is-the-same-mlst">Choosing a reference that is the same MLST</h4>
<p>If all your samples belong to the same ST, you can choose an appropriate high-quality reference that is also of the same ST.</p>
</section>
<section id="choosing-a-reference-that-is-within-a-pre-determined-mashani-threshold" class="level4">
<h4 class="anchored" data-anchor-id="choosing-a-reference-that-is-within-a-pre-determined-mashani-threshold">Choosing a reference that is within a pre-determined MASH/ANI threshold</h4>
<p>If an appropriate high-quality reference genome is not available, an alternative is to choose one of your own samples as the reference. This can be done by performing all-vs-all ANI or MASH distance estimation, and picking a reference genome that is approximately equidistant from all the remaining genomes (i.e the midpoint or centroid).</p>
</section>
<section id="performing-multiple-independent-alignments" class="level4">
<h4 class="anchored" data-anchor-id="performing-multiple-independent-alignments">Performing multiple independent alignments</h4>
<p>In some cases, your dataset might have two distinct lineages (subspecies/MLST types). In these cases, it is advisable perform your analyses separately with an appropriate reference for each group.</p>
</section>
<section id="using-reference-free-methods" class="level4">
<h4 class="anchored" data-anchor-id="using-reference-free-methods">Using reference-free methods</h4>
<p>If your dataset is highly diverse with mutliple potential lineages, you can perform reference-free SNP calling. For example, the <strong><em>k</em>-mer based</strong> approach, as employed by SNP calling pipeline <a href="https://www.ncbi.nlm.nih.gov/pubmed/25913206">kSNP</a>, involves breaking up genomic data into odd-length <em>k</em>-mers, and then comparing the <em>k</em>-mers to each other, allowing the center nucleotide to vary.</p>
<p>Other popular methods utilize pan-genome pipelines (e.g., <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02090-4">Panaroo</a>, <a href="https://academic.oup.com/gigascience/article/8/10/giz119/5584409">PIRATE</a>, <a href="https://academic.oup.com/bioinformatics/article/31/22/3691/240757">roary</a>) to identify and align orthologous gene clusters within a set of genomes; the resulting multiple sequence alignments can then be concatenated and queried for SNPs.</p>
<p>Overall, reference-free approaches are valuable in situations where bacterial genomes are distant enough that no reference genome is adequately “close” to all the the genomes in the set.</p>
</section>
</section>
<section id="recombination-filtering" class="level3">
<h3 class="anchored" data-anchor-id="recombination-filtering">Recombination filtering</h3>
<p>SNPs which cluster closely together may have been acquired via recombination and/or horizontal gene transfer (HGT), not independent mutations.</p>
<p>From a SNP calling perspective, this will appear as if a large number of SNPs have occurred one after the other in a particular region of the genome, inflating the SNP distance, when it is really a single event that led to a significant change in the genome.</p>
<p>When your sample set comprises of a set of diverse genomes, even if they are all the same ST (i.e your sample set comprises all the diversity within that ST, and <strong>not</strong> a group of almost identical clones), you must perform recombinaion/HGT masking.</p>
</section>
<section id="core-genome-vs-whole-genome-alignments" class="level3">
<h3 class="anchored" data-anchor-id="core-genome-vs-whole-genome-alignments">Core genome vs whole genome alignments</h3>
<p><b>Genome alignment</b> approaches can be categorized as either core-genome alignment or whole-genome alignment based. Core-genome alignment based approaches involve identifying orthologous sequences present in all genomes in a set, while whole-genome alignment approaches extend alignment to regions not present in all genomes.</p>
<p>In most cases, <strong>core-genome</strong> alignment is the correct approach. For highly clonal samples whole genome alignments can be performed, though in effect this would be equivalent to the core genome alignment.</p>
<p>If your sample set comprises extremely diverse set (eg: multiple species, or several distantly related STs), it is better to perform a concatenated core-gene alignment as done by the pangenome approaches (e.g., <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02090-4">Panaroo</a>, <a href="https://academic.oup.com/gigascience/article/8/10/giz119/5584409">PIRATE</a>)</p>
<p>This approach identifies <strong>core genes</strong> - genes present in all (or most) of your samples, and takes these core gene sequences for each sample and concatenates them together in the same order. This creates an artificial <strong>core genome</strong> for each of your samples and these core genomes are aligned. Pangenomes will be discussed in more detail in a later module.</p>
<section id="exercise-calulate-snp-distances-amongst-a-set-of-genomes" class="level4">
<h4 class="anchored" data-anchor-id="exercise-calulate-snp-distances-amongst-a-set-of-genomes">EXERCISE: Calulate SNP distances amongst a set of genomes</h4>
<p>To calculate SNP distances, we will first perform SNP calling using <code>snippy</code> and obtain a recombination-free <strong>core genome alignment</strong> using <code>gubbins</code>, filter the alignment so that it contains only variable sites using <code>snp-sites</code>, and then calculate SNP distances using <code>snp-dists</code>.</p>
<section id="create-the-snippy-input-file-for-snippy-multi" class="level5">
<h5 class="anchored" data-anchor-id="create-the-snippy-input-file-for-snippy-multi">Create the snippy input file for <code>snippy-multi</code></h5>
<p>To run Snippy <a href="snippy-multi">using multiple genomes as input</a> (i.e., using the <code>snippy-multi</code> command), we need to provide Snippy with a tab-separated input file. This input file will have one line per each of our input genomes, with the name of the sample as the first column and the absolute path to the sample as the second column.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> filename <span class="kw">in</span> <span class="va">$(</span><span class="fu">ls</span> NWU_2025_workshop_data/test_datasets/salmonella_assemblies/<span class="pp">*</span><span class="va">)</span><span class="kw">;</span> <span class="cf">do</span> <span class="va">bname</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$filename</span> _genomic.fna<span class="va">)</span><span class="kw">;</span> <span class="bu">echo</span> <span class="at">-e</span> <span class="st">"</span><span class="va">${bname}</span><span class="st">\t</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span><span class="st">/</span><span class="va">${filename}</span><span class="st">"</span><span class="kw">;</span> <span class="cf">done</span><span class="kw">;</span> <span class="op">&gt;</span> snippy_input.tsv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="run-snippy-multi-to-generate-a-list-of-snippy-commands" class="level5">
<h5 class="anchored" data-anchor-id="run-snippy-multi-to-generate-a-list-of-snippy-commands">Run <code>snippy-multi</code> to generate a list of <code>snippy</code> commands</h5>
<p>Now that we have our input file, we can use <code>snippy-multi</code> to create a script, which will tell Snippy which commands to run. To do this, run the following command:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snippy-multi</span> snippy_input.tsv <span class="at">--ref</span> NWU_2025_workshop_data/test_datasets/reference_genome/genomic.gbff <span class="at">--cpus</span> 1 <span class="op">&gt;</span> runme.sh</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li><code>snippy-multi</code> - Snippy’s command for dealing with multiple genomes as input</li>
<li><code>snippy_input.tsv</code> - Provide <code>snippy-multi</code> with the path to our Snippy input file</li>
<li><code>--ref</code> - Provide <code>snippy-multi</code> with the path to our reference in <code>.gbff</code> format</li>
<li><code>--cpus 1</code> Number of CPUs to use</li>
<li><code>&gt; runme.sh</code> Use <code>stdout</code> redirection with <code>&gt;</code> to direct the output of <code>snippy-multi</code> into a script called <code>runme.sh</code></li>
</ul>
<p>Once <code>snippy-multi</code> finishes, we should see a new file, <code>runme.sh</code>, in our current directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> runme.sh</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>To run Snippy, run the script you just created:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sh</span> ./runme.sh</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This will take some time. Once your Snippy script finishes, you can type <code>ls core.*</code> to see the new files Snippy created:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> core.<span class="pp">*</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>We will be using several of these output files in subsequent steps.</p>
<p>In addition to these output files, Snippy also creates an output directory for each isolate in our data set. We won’t be using these. Let’s create a directory named <code>snippy_final</code> with only the files we are interested in for now:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> snippy_final</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li>Once we’ve created our <code>snippy_final</code> directory, let’s move all of the Snippy results that we want to keep to <code>snippy_final</code> (i.e., all the files starting with <code>core.*</code>):</li>
</ul>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> core.<span class="pp">*</span> snippy_final</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>If we type <code>ls snippy_final</code>, we should see several files within <code>snippy_final</code>, each starting with <code>core.</code>:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> snippy_final</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>For the rest of the lesson, let’s move to our <code>snippy_final</code> directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> snippy_final</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>If we type <code>pwd</code>, we should see that we are in our <code>snippy_final</code> directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="bu">pwd</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>If we type <code>ls</code>, we should see our Snippy output files:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="clean-a-genomic-alignment-produced-using-snippy" class="level5">
<h5 class="anchored" data-anchor-id="clean-a-genomic-alignment-produced-using-snippy">Clean a genomic alignment produced using <code>snippy</code></h5>
<p>The file named <code>core.full.aln</code> is an output file produced by Snippy, which contains our genomic alignment. Specifically, <code>core.full.aln</code> file is a FASTA-formatted multiple sequence alignment file. It has one sequence for the reference, and one for each sample; each sequence has the same length as the reference sequence.</p>
<p>To show the names of the sequences in our genomic alignment, run the following command, which prints the sequence headers in our genomic alignment multi-FASTA:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> core.full.aln </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>We should see all of our input genomes, plus our reference genome (<code>Reference</code>).</p>
<p>We want to supply our genomic alignment to Gubbins so that we can identify and remove recombination. However, <a href="https://github.com/tseemann/snippy"><code>core.full.aln</code> contains some weird characters, which Gubbins (and other downstream programs) may not like.</a> Before we can do anything else with <code>core.full.aln</code>, we need to “clean” it using Snippy’s <code>snippy-clean_full_aln</code> function. This will replace all the “weird” characters with <code>N</code>, so that we can use our genomic alignment as input to other tools.</p>
<p>To clean <code>core.full.aln</code>, let’s run <code>snippy-clean_full_aln</code> and output our clean alignment to a file named <code>clean.full.aln</code>:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snippy-clean_full_aln</span> core.full.aln <span class="op">&gt;</span> clean.full.aln</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Once <code>snippy-clean_full_aln</code> finishes, we can type <code>ls</code> and see that a new file, <code>clean.full.aln</code> has been created:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Our new file, <code>clean.full.aln</code>, contains our clean genomic alignment. We can see that it contains the same sequence headers as <code>core.full.aln</code>:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> clean.full.aln</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="find-and-remove-recombination-in-a-genomic-alignment-using-gubbins" class="level5">
<h5 class="anchored" data-anchor-id="find-and-remove-recombination-in-a-genomic-alignment-using-gubbins">Find and remove recombination in a genomic alignment using <code>gubbins</code></h5>
<p>Now that we have our clean genomic alignment (<code>clean.full.aln</code>), we can use <a href="https://github.com/nickjcroucher/gubbins">Gubbins</a> to detect and remove recombination.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">run_gubbins.py</span> <span class="at">--help</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Let’s use Gubbins to detect and remove recombination in our clean genomic alignment, <code>clean.full.aln</code>; to do this, run the following command:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">run_gubbins.py</span> <span class="at">--prefix</span> gubbins <span class="at">--threads</span> 1 clean.full.aln</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li><code>run_gubbins.py</code> : Call Gubbins</li>
<li><code>--prefix gubbins</code>: Use <code>gubbins</code> as a prefix for the final output filenames</li>
<li><code>--threads 1</code> : Number of threads/CPUs to use</li>
<li><code>clean.full.aln</code> : Path to our clean genomic alignment, which we will be used as input</li>
</ul>
<p>You should see several new files produced with the prefix <code>gubbins.*</code>. You can read all about the output files produced by Gubbins <a href="https://github.com/nickjcroucher/gubbins/blob/master/docs/gubbins_manual.md">here, in the Gubbins manual.</a></p>
<p>We will be using the file <code>gubbins.filtered_polymorphic_sites.fasta</code> in subsequent steps; this file is a multi-FASTA alignment of recombination-filtered SNPs (i.e., it is an alignment of SNPs alone and does not include constant sites).</p>
</section>
<section id="construct-a-recombination-free-core-snp-alignment-using-snp-sites" class="level5">
<h5 class="anchored" data-anchor-id="construct-a-recombination-free-core-snp-alignment-using-snp-sites">Construct a recombination-free core SNP alignment using <code>snp-sites</code></h5>
<p>Now that we have an alignment of recombination-free SNPs present in our genomes, let’s filter that alignment to get only the core SNPs. To do this, we will use <a href="https://github.com/sanger-pathogens/snp-sites">SNP-sites</a>.</p>
<p>Use SNP-sites to construct a final, recombination-free alignment of core SNPs</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snp-sites</span> <span class="at">-c</span> gubbins.filtered_polymorphic_sites.fasta <span class="op">&gt;</span> clean.core.fasta</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li><code>snp-sites</code>: Call SNP-sites</li>
<li><code>-c</code> : An option that tells <code>snp-sites</code> to only output columns containing exclusively ACGT (i.e., core SNPs; )</li>
<li><code>gubbins.filtered_polymorphic_sites.fasta</code>: Ppath to the recombination-free alignment of SNPs produced via <code>gubbins</code> as input</li>
<li><code>&gt; clean.core.fasta</code>: Write our final, recombination-free alignment of core SNPs to a file named <code>clean.core.fasta</code></li>
</ul>
</section>
<section id="calculate-snp-distances-based-on-the-recombination-free-alignment-of-core-snps-using-snp-dists" class="level5">
<h5 class="anchored" data-anchor-id="calculate-snp-distances-based-on-the-recombination-free-alignment-of-core-snps-using-snp-dists">Calculate SNP distances based on the recombination-free alignment of core SNPs using <code>snp-dists</code></h5>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snp-dists</span> <span class="at">-m</span> clean.core.fasta <span class="op">&gt;</span> clean.core.snp-dists.tsv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<ul>
<li><code>snp-dists</code>: Call SNP-dists</li>
<li><code>-m</code> : An option to produce the output in “molten” format. Run the command without this option to see the difference!</li>
<li><code>&gt; clean.core.snp-dists.tsv</code>: Write our SNP distances output to a file called <code>clean.core.snp-dists.tsv</code></li>
</ul>
<hr>
<p>Genomic distance estimation methods have different strengths and applications, and there is no one-size-fits-all approach. While MLST is more for typing strains based on predefined genes, genome distance methods like MASH and ANI can provide broader, more comprehensive comparisons across whole genomes.</p>


</section>
</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>